From 0f4d397c91155f00a9100fe539c5081f3db69a1f Mon Sep 17 00:00:00 2001
From: Gabor Juhos <juhosg@openwrt.org>
Date: Tue, 2 Jun 2026 23:34:12 +0200
Subject: [PATCH] generic: add detach callback to struct phy_driver

lede-commit: fe61fc2d7d0b3fb348b502f68f98243b3ddf5867

Needed by OpenWrt switch layer.

Signed-off-by: Gabor Juhos <juhosg@openwrt.org>
---
 drivers/net/phy/phy_device.c | 3 +++
 include/linux/phy.h          | 6 ++++++
 2 files changed, 9 insertions(+)

diff --git a/drivers/net/phy/phy_device.c b/drivers/net/phy/phy_device.c
index 0615228459ef4..c6522b34d7a8d 100644
--- a/drivers/net/phy/phy_device.c
+++ b/drivers/net/phy/phy_device.c
@@ -1924,6 +1924,9 @@ void phy_detach(struct phy_device *phydev)
 		phydev->devlink = NULL;
 	}
 
+	if (phydev->drv && phydev->drv->detach)
+		phydev->drv->detach(phydev);
+
 	if (phydev->sysfs_links) {
 		if (dev)
 			sysfs_remove_link(&dev->dev.kobj, "phydev");
diff --git a/include/linux/phy.h b/include/linux/phy.h
index fc680901275b7..0e63aaafd4a66 100644
--- a/include/linux/phy.h
+++ b/include/linux/phy.h
@@ -1214,6 +1214,12 @@ struct phy_driver {
 	/** @handle_interrupt: Override default interrupt handling */
 	irqreturn_t (*handle_interrupt)(struct phy_device *phydev);
 
+	/*
+	 * Called before an ethernet device is detached
+	 * from the PHY.
+	 */
+	void (*detach)(struct phy_device *phydev);
+
 	/** @remove: Clears up any memory if needed */
 	void (*remove)(struct phy_device *phydev);
 
